prepareCheckCache function: Add MkdirAll() to restic-cache directory if it doesn't exist

plant99 opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Earlier, if the specified cache dir didn't exist, restic continued `check` command with cache disabled.

Henceforth, it tries to create the cache directory.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------


<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

It was previously discussed here: https://github.com/restic/restic/issues/4437

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].
-->

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

MichaelEischer commented on changelog/unreleased/issue-4437:
```diff
@@ -0,0 +1,7 @@
+Enhancement: In the `check` command, create the cache dir if it doesn't exist.
```
Please change this to
 ```
Enhancement: `check` command creates cache directory if it does not exist
```

MichaelEischer commented on changelog/unreleased/issue-4437:
```diff
@@ -0,0 +1,7 @@
+Enhancement: In the `check` command, create the cache dir if it doesn't exist.
+
+In the `restic check` command, if `CACHE_DIR` is specified but it doesn't exist, check continued with cache disabled.
```
What about something like the following?
```
If a custom cache directory was specified for the `check` command but the directory did not exist,
then `check` continued with cache disabled.

The  `check` command now attempts to create the cache directory before initializing the cache.
```

MichaelEischer commented on cmd/restic/cmd_check.go:
```diff
@@ -172,25 +172,32 @@ func prepareCheckCache(opts CheckOptions, gopts *GlobalOptions) (cleanup func())
 	if err != nil {
-		// if an error occurs, don't use any cache
-		Warnf("unable to create temporary directory for cache during check, disabling cache: %v\n", err)
+		Warnf("unable to create cachedir %s, disabling cache: %vn", cachedir, err)
```
Typo: `"unable to create cache directory %s, disabling cache: %v\n"`

MichaelEischer commented on cmd/restic/cmd_check.go:
```diff
@@ -172,25 +172,32 @@ func prepareCheckCache(opts CheckOptions, gopts *GlobalOptions) (cleanup func())
 		gopts.NoCache = true
 		return cleanup
-	}
+	} else {
```
Instead of `if a { return x; } else { B; }` please write `if a { return x; } B;` (the linter will also complain)

MichaelEischer commented on cmd/restic/cmd_check_test.go:
```diff
@@ -163,3 +165,75 @@ func TestSelectNoRandomPacksByFileSize(t *testing.T) {
+
+func TestPrepareCheckCache(t *testing.T) {
+	// Create a temporary directory for the cache
+	tmpDirBase, err := os.MkdirTemp("", "test-prepare-check-cache-basepath")
```
Please use `t.TempDir()` to get a temporary folder for use in the tests. That tempdir is also automatically removed by the go test framework in the end.

MichaelEischer commented on cmd/restic/cmd_check_test.go:
```diff
@@ -163,3 +165,75 @@ func TestSelectNoRandomPacksByFileSize(t *testing.T) {
+				// listing all directories inside tmpDirBase (cacheDir)
+				// one directory should be tmpDir created by prepareCheckCache with 'restic-check-cache-' in path
+				files, err := os.ReadDir(tmpDirBase)
+				if err != nil {
```
please use `rtest.OK(t, err)` to check for non-nil errors

MichaelEischer commented on cmd/restic/cmd_check_test.go:
```diff
@@ -163,3 +165,75 @@ func TestSelectNoRandomPacksByFileSize(t *testing.T) {
+				if err != nil {
+					t.Errorf("Failed to list cacheDir.")
+				}
+				if len(files) != 0 {
```
Please use
`rtest.Assert(t, len(files) == 0, "expected cache directory not to exist, but it does: %v", files)`. That also applies to the other `len(...) != 0` checks.

plant99 replied:
Is this only for consistency @MichaelEischer , I'm curious if there are more benefits than ```len(...)``` checks?

MichaelEischer replied:
The main benefit is that `rtest.Assert` takes less space to write than the `if ... { t.Errof(...) }`. By now, there's also `rtest.Equals(t, 0, len(files), "description")` that automatically prints both values if there's a mismatch.

MichaelEischer commented on cmd/restic/cmd_check_test.go:
```diff
@@ -163,3 +165,75 @@ func TestSelectNoRandomPacksByFileSize(t *testing.T) {
+				// If using a temporary cache directory, the cache directory should exist
+				// listing all directories inside tmpDirBase (cacheDir)
+				// one directory should be tmpDir created by prepareCheckCache with 'restic-check-cache-' in path
+				files, err := os.ReadDir(tmpDirBase)
```
Both code paths call `os.ReadDir`, that call should be moved out of the if block.

MichaelEischer commented on cmd/restic/cmd_check_test.go:
```diff
@@ -163,3 +165,75 @@ func TestSelectNoRandomPacksByFileSize(t *testing.T) {
+
+	testCases := []struct {
+		opts        CheckOptions
+		cacheExists bool
```
`cacheExists` and `expectNoErr` are unused.

MichaelEischer commented on cmd/restic/cmd_check_test.go:
```diff
@@ -163,3 +165,75 @@ func TestSelectNoRandomPacksByFileSize(t *testing.T) {
+		expectNoErr bool
+	}{
+		{CheckOptions{WithCache: true}, false, true}, // Shouldn't create temp directory
+		{CheckOptions{WithCache: false}, true, true}, // Should create temp directory
```
Those tests don't exercise the new code paths as `tmpDirBase` already exists.

plant99 replied:
I have introduced a new flag called `withValidCache` which if false removes the `tmpDirBase`, and tries the new code path. I hope that seems good enough.

MichaelEischer commented on cmd/restic/cmd_check_test.go:
```diff
@@ -163,3 +165,75 @@ func TestSelectNoRandomPacksByFileSize(t *testing.T) {
+			if len(files) != 0 {
+				t.Errorf("Expected cache directory not to exist, but it does")
+			}
+			if len(files) != 0 {
```
The check is duplicated.

MichaelEischer reviewed (changes requested):
Thanks for working on this issue! The code change itself is mostly fine, except for two minor issues. 

However, the test case only seems to test existing behavior, but not the cache directory creating behavior added by this PR. Please fix that.

In addition, the PR must also be rebased, we don't want to merge merge commits. Ideally also squash the commits.

plant99 committed:
check: enhance check command to create cache directory if it doesn't exist

plant99 commented:
@MichaelEischer 
Thanks for the detailed review of this PR :) , I've addressed your concerns.

Please review the following two things again:
 - the single commit message
 - the new test iteration which checks the `MkdirAll` code paths.

MichaelEischer reviewed (approved):
LGTM. Thanks!
